### ISSUE TITLE
Signatur immer anders - rk MOBILE

### ISSUE THREAD

**Issue by nn1234** [a-sit-plus/at-registrierkassen-mustercode#410]
Hallo,

ist es normal das die Signatur jedesmal anders ist bei gleichem Payload. Ist das normal?

**Soll signiert werden**
_R1-AT1_MY-Bookings-01_1_2017-03-11T16:34:09_0,00_0,00_0,00_0,00_0,00_7L5qAhxzzfo=_62A71C38_8WBwOaOEr8g=

**Ergebnis 1:**
    [header] => eyJhbGciOiJFUzI1NiJ9
    [payload] => X1IxLUFUMV9NWS1Cb29raW5ncy0wMV8xXzIwMTctMDMtMTFUMTY6MzQ6MDlfMCwwMF8wLDAwXzAsMDBfMCwwMF8wLDAwXzdMNXFBaHh6emZvPV82MkE3MUMzOF84V0J3T2FPRXI4Zz0
    [**signature**] => NJstLsefibbFqdGQBxx-0p-dSZpv41DjpqEgW-aLva-YJJzdGLE1KE-XVzKOoDRlHL5eBGbJP-I43zKgt3BN0A

**Ergebnis 2:**
    [header] => eyJhbGciOiJFUzI1NiJ9
    [payload] => X1IxLUFUMV9NWS1Cb29raW5ncy0wMV8xXzIwMTctMDMtMTFUMTY6MzU6MjFfMCwwMF8wLDAwXzAsMDBfMCwwMF8wLDAwXzdMNXFBaHh6emZvPV82MkE3MUMzOF84V0J3T2FPRXI4Zz0
    [**signature**] => e-kwSFmZ_Bq0pmYjUIMXssGQ1RzGai3RpB93umw1DWrv4zvIraS4ww4i4IuH0xAo2YFtICRw9YBdJqUanbI6bw

**Comment by ztp-mino** [comment]
Ist normal. IIRC ist es notwendig einen Zufallswert mit der Signatur zu verwenden weil sich sonst mit ein paar mathematischen Tricks aus einer Signatur Signaturen zu anderen Daten anfertigen lassen.

**Comment by nn1234** [comment]
Danke

**Comment by andreasschiestl** [comment]
@daJedi Auch bei komplett identischer Payload würden sich unterschiedliche Signaturen ergeben aufgrund des von ztp-mino erwähnten Zufallswert der für die Signaturerzeugung verwendet wird. Ganz so einfach ist es aber nicht da die Signatur im Fall von ES256 aus dem Hashwert der Daten berechnet wird und der Sinn einer kryptographische Hashfunktion (wie SHA-256 es ist) ist es das zu verhindert bzw. erheblich zu erschweren.

**Comment by ztp-mino** [comment]
Ja, hab mich da vertan, das war bei RSA. Bei dem verwendeten EC DSA ist es sogar noch tragischer. Wenn kein Zufallswert benutzt wird, bzw. für zwei Payloads der gleiche Wert benutzt wird, lässt sich der Private Key rekonstruieren.

**Comment by daJedi** [comment]
@andreasschiestl das wurde bereits von meinem Vorposter beantwortet und daher wollte ich nur noch auf den Fehler  im 1. Satz von @nn1234 hinweisen ... ohne Zufallszahlen macht keine Crypto Sinn, oder? ;-)

**Comment by andreasschiestl** [comment]
@daJedi wollte es nur noch mal betonen, dass es eben auch bei identischer Payload nicht zur gleichen Signatur kommt. Viele die von der RKSV betroffen sind haben wenig bis gar nichts mit Kryptographie zu tun :)
